### DISCUSSION TITLE
Many-to-many relationship with dataclasses causes recursion error with jsonify

### DISCUSSION THREAD

**Discussion by Makistos** [sqlalchemy/sqlalchemy#10366] [Usage Questions] [answered]
I can't seem to be able to run jsonify when I have a many-to-many relationship. Somewhere someone said this should be solvable by setting repr to False, but that did not help. How should I set my relationships up so I avoid this issue? This is what it looks (don't ask about the table names, I did not create this db):
```
class Base(MappedAsDataclass, DeclarativeBase):
    pass

class CommonMixin():
    '''Common table definitions

    Returns:
        _type_: _description_
    '''
    @declared_attr.directive
    def __tablename__(cls) -> str:
        return cls.__name__.lower()  # type: ignore
    __table_args__ = {"mysql_engine": "InnoDB"}


class Trackable(CommonMixin):
    ''' Mixin for tables that have id, created_at and updated_at columns.
    '''
    id: Mapped[int] = mapped_column(primary_key=True)
    created_at = mapped_column(DateTime, server_default=func.now())
    updated_at = mapped_column(DateTime, onupdate=func.now())

attachment_sets_attachments = Table(
    "attachment_sets_attachments",
    Base.metadata,
    Column("attachment_set_id", Integer, ForeignKey(
        "attachment_sets.id"), primary_key=True),
    Column("attachment_id", Integer, ForeignKey(
        "attachments.id"), primary_key=True)
)


class attachment_sets(Trackable, Base):
    host_id: Mapped[int] = mapped_column()
    host_type: Mapped[str] = mapped_column(String(255))
    host_version: Mapped[int] = mapped_column(Integer)
    attachments: Mapped[List["attachments"]] = relationship(
        "attachments", secondary=attachment_sets_attachments, viewonly=True,
        repr=False, backref='attachments.id')

class attachments(Trackable, Base):
    orig_filename: Mapped[str] = mapped_column(String(255))
    type: Mapped[str] = mapped_column(String(255))
    data: Mapped[str] = mapped_column(Text)
    attachment_sets: Mapped[List["attachment_sets"]] = relationship(
        "attachment_sets", secondary=attachment_sets_attachments, viewonly=True,
        repr=False, backref='attachment_sets.id')
```

**Comment by CaselIT** [comment]
Hi,

I don't know what `jsonify` is, but it seems it does not supports structures that have cycles in them.
This is not an sqlalchemy issue.

As for solution:
- use a different serializer that takes into consideration cycles
- don't load the relationships when loading the data, by using `defer` or similar options
- tell the serializer what fields to serialize, so you can exclude the cycle

**Reply by Makistos** [reply]
Jsonify is just a wrapper around json.dumps() function: https://tedboy.github.io/flask/generated/flask.jsonify.html. json.dumps() gives the same issue. 

What serializer would work in Python world in this case? I have used marshmallow, but it's awfully slow and requires me to write a separate schema for it instead of using dataclasses. I'm trying to make it work like defined here:  https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping. Using marshmallow would defeat the whole idea of dataclasses.

**Reply by zzzeek** [reply, marked as answer]
you can either not use bi-directional relationships, or do it the old fashioned way and add serialization functions (this is what I would do personally, I'd likely have it use inspect(mapper).attrs to get the list of names I want to serialize and I'd programmatically set up how I want it to look).

```py
class MyClass:
    def serialize(self):
        return {
            "name": self.name,
            "related_things": [thing.serialize() for thing in self.things]
        }
```

**Reply by CaselIT** [reply]
> What serializer would work in Python world in this case?

As mike suggested, I usually go with a custom serialization function.
In general you look for a serializer that can support dataclasses with cycles in them, or at least at one where you can tell not to serialize some attributes.
`dataclasses.asdict` does not seem to support cycles either

**Reply by Makistos** [reply]
Okay, this sounds like a good idea, especially as I shouldn't have too many of these circular issues. So mostly I can use existing serializers.
